File: functional-programming/src/main/scala/Phonecode.scala

/** 
 * reaktor dev day, M. Odersky
 */
class Coder(words: List[String]) {

  private val mnemonics = Map(
      '2' -> "ABC", '3' -> "DEF", '4' -> "GHI", '5' -> "JKL", 
      '6' -> "MNO", '7' -> "PQRS", '8' -> "TUV", '9' -> "WXYZ")
      
  private val charCode: Map[Char, Char] =
    for ((digit, str) <- mnemonics; ltr <- str) yield (ltr -> digit)
          
  private def wordCode(word: String): String = word.toUpperCase map charCode
    
  private val wordsForNum: Map[String, Seq[String]] = (words groupBy wordCode) withDefaultValue List()
  
  def encode(number: String): List[List[String]] = 
    if (number.isEmpty()) List(List())
    else {
      for {
        split <- 1 to number.length
        word <- wordsForNum(number take split)
        rest <- encode(number drop split)
      } yield word :: rest
    }.toList
    
}

object Phonecode extends App {
  val start = System.currentTimeMillis()
  val dict = io.Source.fromFile("../haskell/testwords")
     .getLines.filter(_.length > 1).filter(_.matches("[a-zA-Z]+")).toList  
  val coder = new Coder(dict)
  println("Using " + dict.length + " words.")
  println("Found " + coder.encode("222667542489").length + " solutions.")
  val stop = System.currentTimeMillis()
  stop - start
  println("Time " + ((stop - start)/1000.0) + "s")
}

File: functional-programming/src/main/haskell/Phonecode.hs

import Prelude hiding (lookup)
import Data.Map hiding (map, filter)
import Data.Maybe
import Data.Char
import Data.Time.Clock

mnemonics = fromList [ ('2', "ABC"), ('3', "DEF"), ('4', "GHI"), ('5', "JKL")
                     , ('6', "MNO"), ('7', "PQRS"), ('8', "TUV"), ('9', "WXYZ") ]

get key dict = fromJust (lookup key dict)

charCode = fromList [(letter, digit) | digit <- keys mnemonics, letter <- get digit mnemonics]

wordCode = map (flip get charCode . toUpper) 

wordsForNum (word:[]) = fromList [(wordCode word, [word])]
wordsForNum (word:words) = insertWith (++) (wordCode word) [word] (wordsForNum words)

encode wordList "" = [[]]
encode wordList code = [ word : rest | i <- [1.. length code]
                                     , let words = lookup (take i code) (wordsForNum wordList) 
                                     , words /= Nothing
                                     , word <- fromJust words 
                                     , rest <- encode wordList (drop i code) ]

main = do 
	t1 <- getCurrentTime
	str <- readFile "testwords"
	let words = filter (and . map isAlpha) $ filter ((>1) . length) $ lines str
	let solutions = encode words "222667542489"
	putStrLn $ "Using dictionary with " ++ show (length words) ++ " words." 
	putStrLn $ "Counted " ++ show (length solutions) ++ " solutions."
	t2 <- getCurrentTime
	putStrLn $ "Time: " ++ show (diffUTCTime t2  t1)
	